### PR TITLE
Run the update-staging-godeps script inside a docker container

### LINKED ISSUE #45757: update-staging-godeps.sh should not depend on local Go version
In https://github.com/kubernetes/kubernetes/pull/45752 it was discovered that `update-staging-godeps.sh` generates different output depending on your local Go version, rather than pinning the Go version inside a Docker build container.

This makes it dangerous to run `update-all.sh` on a release branch, since you may have a different Go version installed than the one on which the release branch is based.

CC @caesarxuchao @sttts

### PR DESCRIPTION
**Pull request by cblecker**
**What this PR does / why we need it**:
This PR moves the update-staging-godeps script to run inside a docker container.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #45757 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

/assign @ixdy @sttts

### CODE REVIEW

**Comment by cblecker** [comment]
This is still a work in progress, but I wanted to get a PR out there for an initial pass as far as the right direction.

I've run into a problem that I haven't figured out yet though.. I'm getting odd errors from godep that I haven't seen before:
```
Checking whether godeps are restored
Updating godeps for k8s.io/apimachinery
Rewriting Godeps.json to remove commits that don't really exist because we haven't pushed the prereqs yet
Updating godeps for k8s.io/apiserver
  godep: error while inspecting "/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/discovery": directory "/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/discovery" is not using a known version control system
  godep: error while inspecting "/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/discovery/fake": directory "/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/discovery/fake" is not using a known version control system
  godep: error while inspecting "/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/informers": directory "/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/informers" is not using a known version control system
  godep: error while inspecting "/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/informers/apps": directory "/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/informers/apps" is not using a known version control system
[...snip...]
```

If you have any leads @ixdy or @sttts, please let me know! Otherwise, I'll kick it more next week and see what I can figure out.

**Comment by sttts** [comment]
@cblecker that's related to line https://github.com/kubernetes/kubernetes/pull/46193/files#diff-e87faa4ff683dd595670fa1278c5a066R69 and that godep does not find the .git directory in any of the parent dirs of `/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/discovery`. Not sure about the rsyncd change. I guess you try to sync the .git dir as well? You could also do a `git init` in the clone of the k8s repo, if that's easier or faster.

**Comment by cblecker** [comment]
@sttts See, that's what I thought too, but I am rsyncing the .git dir in now. If I `build/shell.sh` in, here's what I see:
```
I have no name!@kube-build:~$ cd /go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/discovery
I have no name!@kube-build:~/staging/src/k8s.io/client-go/discovery$ git rev-parse --git-dir
/go/src/k8s.io/kubernetes/.git
I have no name!@kube-build:~/staging/src/k8s.io/client-go/discovery$ git status
On branch staging-godeps-dockerized
Your branch is up-to-date with 'origin/staging-godeps-dockerized'.
nothing to commit, working directory clean
I have no name!@kube-build:~/staging/src/k8s.io/client-go/discovery$ 
```

**Comment by sttts** [comment]
@cblecker godep is searching the .git dir. It has some (broken) logic to select the `$GOPATH` an absolute paths belongs to. E.g. if it finds a dependency in the 3rd component of `$GOPATH` (which is `${KUBE_ROOT}/staging`), it actually "assignes" it to the first one in `$GOPATH` which is a prefix. This is `/go`, not `${KUBE_ROOT}/staging`. By this it used to find the `.git` directory, though with a wrong relative path. Not sure this is broken now in your setup.

**Comment by cblecker** [comment]
@sttts @ixdy I think this should be ready for review. I'd personally hold off on pushing the kube-cross image as I know in the next <24h we'll get an image based on go 1.8.3 and I'll rebase this on top of that.. but if you build kube-cross locally, this seems to do the trick.

Thanks for the lead on the gopath error, @sttts. The gopath was being messed with, so it couldn't find the right staging dir. I was able to figure out the permissions issues on the gopath dir in the build image, and now the primary `/go` gopath is mutable

**Comment by cblecker** [comment]
Rebased on the go1.8.3 kube-cross image

**Comment by cblecker** [comment]
@sttts @ixdy Is there a chance of this getting reviewed before code freeze (EOD 6/1)?

File: build/build-image/Dockerfile [hunk 1]
```diff
@@ -36,6 +36,13 @@ ENV KUBE_OUTPUT_SUBPATH _output/dockerized
 # Pick up version stuff here as we don't copy our .git over.
 ENV KUBE_GIT_VERSION_FILE ${HOME}/.dockerized-kube-version-defs
 
+# Add system-wide git user information
+RUN git config --system user.email "nobody@k8s.io" \
+  && git config --system user.name "kube-build-image"
+
+# Fix permissions on gopath
+RUN chmod -R a+rwx $GOPATH
```
Comment (sttts) [inline build/build-image/Dockerfile]: even though this is inside docker, 0777 rings the alarm bell.

File: build/build-image/rsyncd.sh [hunk 2]
```diff
@@ -77,7 +77,7 @@ port = 8730
   secrets file = ${SECRETS}
   read only = false
   path = ${VOLUME}
-  filter = - /.make/ - /.git/ - /_tmp/
+  filter = - /.make/ - /_tmp/
```
Comment (sttts) [inline build/build-image/rsyncd.sh]: .git is 430 MB in my case. Do we really want to sync that into the container?

Comment (sttts) [inline build/build-image/rsyncd.sh]: alternative: shallow git clone.

File: hack/update-staging-godeps-dockerized.sh [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      echo "Invalid option: -$OPTARG" >&2
+      exit 1
+      ;;
+  esac
+done
+
+kube::util::ensure_godep_version v79
+
+echo "Checking whether godeps are restored"
+if ! kube::util::godep_restored 2>&1 | sed 's/^/  /'; then
```
Comment (sttts) [inline hack/update-staging-godeps-dockerized.sh]: do we have caching? I.e. is the GOPATH preserved? I guess yes, if we have the persisted storage container. But please confirm.

Comment (cblecker) [inline hack/update-staging-godeps-dockerized.sh]: I'm not actually persisting the storage here, so no, this doesn't cache the godeps atm.

File: build/build-image/rsyncd.sh (hunk 2 above)
Comment (cblecker) [inline build/build-image/rsyncd.sh]: because it's an rsync and not a clone, I'm not sure it makes a difference. the system running this operation should already have a full copy of the repo, and it doesn't seem to take a long time to sync it into the container.

File: build/build-image/Dockerfile (hunk 1 above)
Comment (cblecker) [inline build/build-image/Dockerfile]: The reason for this is here: https://github.com/kubernetes/kubernetes/pull/46193/files#diff-8188c017a3a4a9343bc9920417e2376bR70

We're `go get`ing these tools as root when the kube-cross container is built, which breaks the perms on the GOPATH when we try to restore godeps

File: build/build-image/rsyncd.sh (hunk 2 above)
Comment (sttts) [inline build/build-image/rsyncd.sh]: it's not that slow, I tried today

Comment (cblecker) [inline build/build-image/rsyncd.sh]: okay, so syncing it into the container is okay with you?

File: hack/update-staging-godeps-dockerized.sh (hunk 3 above)
Comment (cblecker) [inline hack/update-staging-godeps-dockerized.sh]: let me look into this

File: build/build-image/rsyncd.sh (hunk 2 above)
Comment (ixdy) [inline build/build-image/rsyncd.sh]: I think we need `.git` as otherwise `hack/lib/version.sh` won't work.

File: build/common.sh [hunk 4]
```diff
@@ -672,14 +672,10 @@ function kube::build::sync_to_container() {
   # they will not be re-generated by 'make'.
   kube::build::rsync \
     --delete \
-    --filter='+ /staging/**' \
-    --filter='- /.git/' \
     --filter='- /.make/' \
     --filter='- /_tmp/' \
     --filter='- /_output/' \
     --filter='- /' \
-    --filter='- zz_generated.*' \
```
Comment (ixdy) [inline build/common.sh]: why did you delete these?

File: hack/update-staging-godeps-dockerized.sh [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+# TODO this does not address client-go, since it takes a different approach to vendoring
+# TODO client-go should probably be made consistent
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+source "${KUBE_ROOT}/hack/lib/init.sh"
+source "${KUBE_ROOT}/hack/lib/util.sh"
```
Comment (ixdy) [inline hack/update-staging-godeps-dockerized.sh]: nit: `hack/lib/init.sh` includes util.sh, so this isn't needed.

File: hack/update-staging-godeps-dockerized.sh [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+      exit 1
+      ;;
+  esac
+done
+
+kube::util::ensure_godep_version v79
+
+echo "Checking whether godeps are restored"
+if ! kube::util::godep_restored 2>&1 | sed 's/^/  /'; then
+  hack/godep-restore.sh
```
Comment (ixdy) [inline hack/update-staging-godeps-dockerized.sh]: maybe `${KUBE_ROOT}/hack/godep-restore.sh`?

File: hack/update-staging-godeps-dockerized.sh [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+  updateGodepManifest "${repo}"
+
+  if [ "${FAIL_ON_DIFF}" == true ]; then
+    diff --ignore-matching-lines='^\s*\"GoVersion\":' --ignore-matching-line='^\s*\"GodepVersion\":' --ignore-matching-lines='^\s*\"Comment\"' -u "${KUBE_ROOT}/staging/src/k8s.io/${repo}/Godeps" "${TMP_GOPATH}/src/k8s.io/${repo}/Godeps/Godeps.json"
+  fi
+  if [ "${DRY_RUN}" != true ]; then
+    cp "${TMP_GOPATH}/src/k8s.io/${repo}/Godeps/Godeps.json" "${KUBE_ROOT}/staging/src/k8s.io/${repo}/Godeps"
+    # Assume Godeps.json is not updated, so that it doesn't make the working tree dirty
+    git update-index --assume-unchanged "${KUBE_ROOT}/staging/src/k8s.io/${repo}/Godeps/Godeps.json"
```
Comment (ixdy) [inline hack/update-staging-godeps-dockerized.sh]: it looks like this was newly added, not copied from the original, but I'm having trouble understanding why we need it.

File: hack/update-staging-godeps.sh [hunk 8]
```diff
@@ -14,98 +14,20 @@
...
+kube::util::ensure_clean_working_dir
 
-  pushd "${TMP_GOPATH}/src/k8s.io/${repo}" >/dev/null
-    git init >/dev/null
-    git config --local user.email "nobody@k8s.io"
-    git config --local user.name "$0"
-    git add . >/dev/null
-    git commit -q -m "Snapshot" >/dev/null
-  popd >/dev/null
+# NOTE: All output from this script needs to be copied back to the calling
```
Comment (ixdy) [inline hack/update-staging-godeps.sh]: I kinda wish we only copied output back if `DRY_RUN=false` (which would eliminate all of the `DRY_RUN` logic from `update-staging-godeps-dockerized.sh`), but I also recognize that that would make the `build/*` scripts more complicated.

(no action needed, just a comment.)

**Review by ixdy — COMMENTED** [review]
I'm mostly ignoring `update-staging-godeps-dockerized.sh` since it appears to largely be a copy of the non-dockerized version.

File: hack/update-staging-godeps-dockerized.sh (hunk 7 above)
Comment (cblecker) [inline hack/update-staging-godeps-dockerized.sh]: The godeps script requires a clean git tree. Right now, it requires you commit each file, and pauses on each repo in the run that has changes (it's a real PITA). What this does is ignores changes to the file we know is changing, but the git tree inside the container doesn't matter, as we rsync the changed files outs.

File: hack/update-staging-godeps.sh (hunk 8 above)
Comment (cblecker) [inline hack/update-staging-godeps.sh]: I'll think about this when I crack this open again for another pass and see if there's anything I can do to improve that logic.

File: build/build-image/rsyncd.sh (hunk 2 above)
Comment (sttts) [inline build/build-image/rsyncd.sh]: @cblecker yes.

Comment (ixdy) [inline build/build-image/rsyncd.sh]: also note: this is the same codepath that all dockerized builds use.

File: hack/update-staging-godeps-dockerized.sh (hunk 7 above)
Comment (ixdy) [inline hack/update-staging-godeps-dockerized.sh]: ok. maybe expand the comment a bit to include this explanation?

File: build/build-image/cross/Dockerfile [hunk 9]
```diff
@@ -62,14 +62,13 @@ RUN echo "deb http://archive.ubuntu.com/ubuntu xenial main universe" > /etc/apt/
...
-# Get the code coverage tool, goimports, and godep
 RUN mkdir $TMPDIR \
   && chmod a+rwx $TMPDIR \
-  && chmod o+t $TMPDIR \
-  && go get golang.org/x/tools/cmd/cover \
-            golang.org/x/tools/cmd/goimports \
-            github.com/tools/godep
+  && chmod o+t $TMPDIR
+
+# Get the code coverage tool and goimports
```
Comment (cblecker) [inline build/build-image/cross/Dockerfile]: @ixdy Do we still need these? It would be great if we didn't have to `go get` them.

File: build/common.sh (hunk 4 above)
Comment (cblecker) [inline build/common.sh]: Because I need a clean repo, and if we don't sync in these generated files, the repo will show them as missing

File: build/build-image/cross/Dockerfile (hunk 9 above)
Comment (ixdy) [inline build/build-image/cross/Dockerfile]: it looks like a few generated code tools use `goimports` still.

`hack/make-rules/test.sh` has a coverage mode. I'm not sure how many people still use it. I really don't know how many people use it in combination with the build image.

**Comment by cblecker** [comment]
@ixdy @sttts addressed a bunch of comments, including caching the godeps (🎉). PTAL

I also think the kube-cross changes are set in stone enough (and fairly minor) that a new kube-cross image could be generated and pushed so that CI can run.

**Comment by cblecker** [comment]
and as soon as I push it, I have to rebase haha.. fixed with the changes from https://github.com/kubernetes/kubernetes/commit/83235f940abb0ff638e8d91769cc19fe6df70c0e

**Comment by cblecker** [comment]
@ixdy @sttts okay. rebased again! PTAL :)

**Comment by cblecker** [comment]
@ixdy Did the new kube-cross get pushed? I don't see it in the registry

File: hack/lib/golang.sh [hunk 10]
```diff
@@ -329,11 +329,11 @@ EOF
 
   local go_version
   go_version=($(go version))
-  if [[ "${go_version[2]}" < "go1.6" && "${go_version[2]}" != "devel" ]]; then
+  if [[ "${go_version[2]}" < "go1.8" && "${go_version[2]}" != "devel" ]]; then
```
Comment (sttts) [inline hack/lib/golang.sh]: `min_go_version=1.8`

File: hack/update-staging-godeps-dockerized.sh [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# updates the godeps.json file in the staging folders to allow clean vendoring
+# based on kubernetes levels.
+# TODO this does not address client-go, since it takes a different approach to vendoring
+# TODO client-go should probably be made consistent
```
Comment (sttts) [inline hack/update-staging-godeps-dockerized.sh]: now with k8s.io/api this should be feasible, I guess.

**Comment by sttts** [comment]
tried to run this and got:
```
Cloning into '/go/src/k8s.io/kubernetes/_output/dockerized/go/src/gopkg.in/yaml.v2'...
fatal: unable to access 'https://gopkg.in/yaml.v2/': server certificate verification failed. CAfile: /etc/ssl/certs/ca-certificates.crt CRLfile: none
```
Is there still some image update missing?

File: hack/update-staging-godeps-dockerized.sh (hunk 11 above)
Comment (cblecker) [inline hack/update-staging-godeps-dockerized.sh]: yeah, this todo doesn't apply anymore. will remove.

File: hack/lib/golang.sh (hunk 10 above)
Comment (cblecker) [inline hack/lib/golang.sh]: added a variable

**Comment by cblecker** [comment]
@sttts I just built fresh and tried again.. seemed to work just fine for me. Is there any other debugs or logs you might have that point to where the issue might be? The certificates file is provided by the Debian image that the golang image (and then kube-cross, and kube-build) is built on.

I pulled down `golang:1.8.3`, and tested some commands:
```
$ docker run -it --rm golang:1.8.3  /bin/bash
root@cb9ceb43288e:/go# go get -d gopkg.in/yaml.v2
root@cb9ceb43288e:/go# ls -l src/gopkg.in/yaml.v2
total 336
-rw-r--r-- 1 root root   560 Jul  4 21:10 LICENSE
-rw-r--r-- 1 root root  1313 Jul  4 21:10 LICENSE.libyaml
-rw-r--r-- 1 root root  2599 Jul  4 21:10 README.md
-rw-r--r-- 1 root root 21246 Jul  4 21:10 apic.go
-rw-r--r-- 1 root root 15448 Jul  4 21:10 decode.go
-rw-r--r-- 1 root root 22139 Jul  4 21:10 decode_test.go
-rw-r--r-- 1 root root 45294 Jul  4 21:10 emitterc.go
-rw-r--r-- 1 root root  7458 Jul  4 21:10 encode.go
-rw-r--r-- 1 root root  9121 Jul  4 21:10 encode_test.go
-rw-r--r-- 1 root root 34761 Jul  4 21:10 parserc.go
-rw-r--r-- 1 root root 11950 Jul  4 21:10 readerc.go
-rw-r--r-- 1 root root  5240 Jul  4 21:10 resolve.go
-rw-r--r-- 1 root root 77261 Jul  4 21:10 scannerc.go
-rw-r--r-- 1 root root  2476 Jul  4 21:10 sorter.go
-rw-r--r-- 1 root root   143 Jul  4 21:10 suite_test.go
-rw-r--r-- 1 root root  2435 Jul  4 21:10 writerc.go
-rw-r--r-- 1 root root  9877 Jul  4 21:10 yaml.go
-rw-r--r-- 1 root root 25409 Jul  4 21:10 yamlh.go
-rw-r--r-- 1 root root  4961 Jul  4 21:10 yamlprivateh.go
root@cb9ceb43288e:/go# git clone https://gopkg.in/yaml.v2/
Cloning into 'yaml.v2'...
remote: Counting objects: 936, done.
remote: Total 936 (delta 0), reused 0 (delta 0), pack-reused 936
Receiving objects: 100% (936/936), 1.17 MiB | 1.01 MiB/s, done.
Resolving deltas: 100% (585/585), done.
Checking connectivity... done.
root@cb9ceb43288e:/go# ls -l yaml.v2
total 336
-rw-r--r-- 1 root root   560 Jul  4 21:11 LICENSE
-rw-r--r-- 1 root root  1313 Jul  4 21:11 LICENSE.libyaml
-rw-r--r-- 1 root root  2599 Jul  4 21:11 README.md
-rw-r--r-- 1 root root 21246 Jul  4 21:11 apic.go
-rw-r--r-- 1 root root 15448 Jul  4 21:11 decode.go
-rw-r--r-- 1 root root 22139 Jul  4 21:11 decode_test.go
-rw-r--r-- 1 root root 45294 Jul  4 21:11 emitterc.go
-rw-r--r-- 1 root root  7458 Jul  4 21:11 encode.go
-rw-r--r-- 1 root root  9121 Jul  4 21:11 encode_test.go
-rw-r--r-- 1 root root 34761 Jul  4 21:11 parserc.go
-rw-r--r-- 1 root root 11950 Jul  4 21:11 readerc.go
-rw-r--r-- 1 root root  5240 Jul  4 21:11 resolve.go
-rw-r--r-- 1 root root 77261 Jul  4 21:11 scannerc.go
-rw-r--r-- 1 root root  2476 Jul  4 21:11 sorter.go
-rw-r--r-- 1 root root   143 Jul  4 21:11 suite_test.go
-rw-r--r-- 1 root root  2435 Jul  4 21:11 writerc.go
-rw-r--r-- 1 root root  9877 Jul  4 21:11 yaml.go
-rw-r--r-- 1 root root 25409 Jul  4 21:11 yamlh.go
-rw-r--r-- 1 root root  4961 Jul  4 21:11 yamlprivateh.go
root@cb9ceb43288e:/go# curl -vvvv https://gopkg.in/yaml.v2/ && echo
* Hostname was NOT found in DNS cache
*   Trying 45.33.37.13...
* Connected to gopkg.in (45.33.37.13) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
* SSLv3, TLS handshake, Server hello (2):
* SSLv3, TLS handshake, CERT (11):
* SSLv3, TLS handshake, Server key exchange (12):
* SSLv3, TLS handshake, Server finished (14):
* SSLv3, TLS handshake, Client key exchange (16):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* Server certificate:
* 	 subject: CN=gopkg.in
* 	 start date: 2017-06-30 09:37:00 GMT
* 	 expire date: 2017-09-28 09:37:00 GMT
* 	 subjectAltName: gopkg.in matched
* 	 issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
* 	 SSL certificate verify ok.
> GET /yaml.v2/ HTTP/1.1
> User-Agent: curl/7.38.0
> Host: gopkg.in
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Date: Tue, 04 Jul 2017 21:11:43 GMT
< Content-Length: 71
< Content-Type: text/plain; charset=utf-8
< 
* Connection #0 to host gopkg.in left intact
Unsupported URL pattern; see the documentation at gopkg.in for details.
root@cb9ceb43288e:/go# 
```

All looked fine with the certificate.

**Comment by sttts** [comment]
Tried again, works fine.

**Comment by cblecker** [comment]
Fantastic! Thanks @sttts!

**Comment by sttts** [comment]
What's missing to get the tests green?

**Comment by cblecker** [comment]
@sttts The next step would be getting @ixdy when he has some time to build and push the new kube-cross to gcr.io/google_containers, then the CI tests can run.

**Comment by sttts** [comment]
:+1: looks good from my side

File: hack/update-staging-godeps-dockerized.sh [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    \?)
+      echo "Invalid option: -$OPTARG" >&2
+      exit 1
+      ;;
+  esac
+done
+
+kube::util::ensure_dockerized
+kube::golang::setup_env
+kube::util::ensure_single_dir_gopath
```
Comment (cblecker) [inline hack/update-staging-godeps-dockerized.sh]: Added this new function. Thanks @sttts!

**Comment by cblecker** [comment]
@ixdy any update on this? :)

**Comment by ixdy** [comment]
sorry for the extended delay on this one. I'm going to go build and push cross-build:v1.8.3-2 now.

**Comment by ixdy** [comment]
pushed.

/retest

**Comment by ixdy** [comment]
eventual consistency issues?
/retest

**Comment by ixdy** [comment]
welp, we need to update the `kubekins-test` image.
from the unit test build:
```
W0713 23:02:11.440] Detected go version: go version go1.8.1 linux/amd64.
W0713 23:02:11.440] Kubernetes requires go version go1.8.3 or greater.
```

sure enough, https://github.com/kubernetes/test-infra/blob/master/jenkins/test-image/Dockerfile-1.7#L18
@kubernetes/test-infra-maintainers anyone want to fix this?

**Comment by krzyzacy** [comment]
@ixdy this is for 1.8 only or it's going to be cherrypick to 1.7? (depend on if I need to bump 1.7 go version or I need to add a 1.8 Dockerfile)

**Comment by ixdy** [comment]
I don't think this PR is being cherry-picked to 1.7, but 1.7 uses go1.8.3, not go1.8.1, so we should probably update the 1.7 Dockerfile.

**Comment by caesarxuchao** [comment]
FYI, we currently have dependency cycle in the code base so it might cause hack/update-staging-godeps.sh malfunctioning. I sent https://github.com/kubernetes/kubernetes/pull/48906 to fix.

**Comment by cblecker** [comment]
cross-build is currently broken: https://github.com/kubernetes/kubernetes/issues/48887

**Comment by cblecker** [comment]
@ixdy Thanks for the cross image push!

@krzyzacy Looks like pull-kubernetes-node-e2e is failing because of version requirements too. Is that the e2e image that @ixdy mentioned here: https://github.com/kubernetes/test-infra/pull/3500#issuecomment-315230429 ? Or should that have been fixed by the kubekins image?

**Comment by krzyzacy** [comment]
ah ack, I'll update the node image as well

**Comment by krzyzacy** [comment]
image pushed, test is actually using latest tag so no code change is needed

**Comment by cblecker** [comment]
@krzyzacy thanks for your help!

@sttts @ixdy All tests pass. Once that unrelated Azure vendoring issue breaking the cross build is fixed, this should be good to go.

**Comment by cblecker** [comment]
@sttts @ixdy Cross build now passing :)

File: hack/verify-staging-godeps.sh [hunk 13]
```diff
@@ -19,4 +19,4 @@ set -o nounset
 set -o pipefail
 
 KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
-${KUBE_ROOT}/hack/update-staging-godeps.sh -d -f "$@"
+KUBE_VERBOSE=3 KUBE_RUN_COPY_OUTPUT=N ${KUBE_ROOT}/hack/update-staging-godeps.sh -d -f "$@"
```
Comment (sttts) [inline hack/verify-staging-godeps.sh]: The KUBE_VERBOSE was only for testing?

Comment (cblecker) [inline hack/verify-staging-godeps.sh]: Nope, this was intended. Based on the other scripts (e.g. https://github.com/kubernetes/kubernetes/blob/master/hack/verify-generated-protobuf.sh#L41) VERBOSE 3 gives the right log output for CI

**Comment by cblecker** [comment]
@ixdy Over to you for approve :)

**Comment by cblecker** [comment]
@sttts @ixdy Rebased to include changes from https://github.com/kubernetes/kubernetes/pull/49095.

**Comment by cblecker** [comment]
Tests are broken until #49247 merges

**Comment by cblecker** [comment]
@sttts @ixdy rebased again. all tests passing. PTAL

**Comment by ixdy** [comment]
/lgtm

gonna assume it all still actually works. I guess at least the verify checks are working. :)

**Comment by krzyzacy** [comment]
huh how did this merge happen with verify failed and gce still in progress?

**Comment by cblecker** [comment]
Merged via batch. Looks like the job verify timed out? It might be because it's downloading godeps multiple times (once outside in container for this job, once at the beginning for the rest of the out of container scripts).

**Comment by CaoShuFeng** [comment]
Hi @cblecker 
I work behind a proxy.
/hack/update-staging-godeps.sh fails because it tried to run "go get" inside a container without proxy set.

**Comment by cblecker** [comment]
@CaoShuFeng Can you open an issue and mention me in it? Shouldn't be a complicated fix, but don't want to forget about it.

**Comment by CaoShuFeng** [comment]
@cblecker 
The issue:
https://github.com/kubernetes/kubernetes/issues/49921

**Comment by ivan4th** [comment]
Just noticed this change, it completely broke my workflow. I use a remote docker engine for building k8s and running [kubeadm-dind-cluster](https://github.com/Mirantis/kubeadm-dind-cluster) there, and my connection speed is limited (around 1 Mb/s), so rsyncing 400+ Mb `.git` directory is a disaster. Is there any chance of make rsyncing `.git` optional at the cost of breaking update scripts if it's turned off? If so, I can create a PR.

**Comment by ixdy** [comment]
Yikes.
Maybe only sync `.git` when running the update scripts, along with printing a warning that it might be slow?

**Comment by cblecker** [comment]
Yeah, that might work @ixdy.. although we'd need to check if switching between the two would invalidate the data container.

@ivan4th Would you mind opening a new issue for this so it can be tracked? If you want to take a crack, be my guest. Otherwise I should be able to take a look in the next week or so to see work involved.

**Comment by ivan4th** [comment]
Created #50272, I'll work on it.
